Initial setup for PXF port into GPDB

denalex opened the pull request:
This PR sets up a framework for porting PXF from Apache HAWQ into GPDB. PXF will allow GPDB to query and write to Hadoop-based datasources.

This PR creates a new extension "pxf" and contains mock code that returns a preset tuple. Unit tests, regression test, and continuous integration jobs are also included. This is done so that we can follow iterative development approach, adding functionality and reducing the level of mocking with each subsequent PR while maintaining green state for tests. We decided that waiting to create a PR until the full functionality is ported would be a worse choice as the amount of changes would be overwhelming. 

Also a new configure option --enable-pxf is introduced to produce builds with PXF.

pivotal-issuemaster commented:
@denalex Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=2634)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=2634) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

d commented on gpAux/extensions/pxf/test/Makefile:
```diff
@@ -0,0 +1,15 @@
+
+include $(top_builddir)/src/backend/mock.mk
+
+#aocsam.t: \
```
😱 
dead code

d commented on gpAux/extensions/pxf/test/pxfprotocol_test.c:
```diff
@@ -0,0 +1,26 @@
+#include <setjmp.h>
+#include "cmockery.h"
+
+//#include "../src/pxfprotocol.c"
```
😱 
dead code / dead comment

d commented on gpAux/extensions/pxf/test/pxfprotocol_test.c:
```diff
@@ -0,0 +1,26 @@
+void
+test_pxfprotocol_export(void **state)
+{
+    //assert_int_equal(pxfprotocol_export(), 0);
```
OMG 😱

d commented on gpAux/extensions/pxf/test/pxfprotocol_test.c:
```diff
@@ -0,0 +1,26 @@
+#include <stdarg.h>
```
This entire file is a pointless existence: it runs no code, tests nothing

denalex replied:
I don't agree -- this is a unit test for whatever functionality pxfprotocol.c currently provides. Specifically, it tests that given a set of data from External Protocol Manager, the import function returns expected data.

hlinnaka replied:
I agree with Jesse, this is pointless. More to the point, this doesn't cover anything that isn't also covered by the installcheck test. CMockery tests are extremely useful to test low-level internal functions, but they're not very well suited for this kind of testing. I would just remove this file.

If CMockery tests become more interesting in the follow-up work, then let's add it at that point.

denalex replied:
@d has commented on somehow outdated version of the file. Our understanding is that installcheck is an integration end-to-end regression framework that issues SQL statements and compares output, while these are unit tests proving that c functions produce expected output given a specific input. We will revisit this file once the next layer "gpbridge" is introduced and will need to get mocked in this test.

d commented on gpAux/extensions/pxf/src/pxfprotocol.c:
```diff
@@ -0,0 +1,104 @@
+Datum pxfprotocol_import(PG_FUNCTION_ARGS);
+Datum pxfprotocol_validate_urls(PG_FUNCTION_ARGS);
+
+typedef struct GpId {
```
Both the `typedef GpId` and `extern` are already declared in `cdb/cdbvars.h`

d commented:
Meh. This patch set seems too small to be a "small step". Specifically, even if the intent is to bring in the feature in batches of small increments, this is too small or too insignificant of a "step one". What does "step one plus step two" look like, then? If that's going to be too big, then step two alone will definitely still be very very big.

denalex commented:
@d -- thanks, a valid question. Definitions of "small" are obviously subjective. This PR defines protocol handler, implements read function with mock data and introduces unit / regression tests along with CI job. This was a planned scope for the first step.

Further steps will port libchurl from HAWQ, introduce gpbridge layer, port data fragmentation logic, bring PXF setup into CI, etc ... Only at the end it will start working end-to-end without mock data, but we want to keep integrating this work as we complete each step, instead of accumulating changes in a long-running feature branch.

@schubert @pivotal-mike @vraghavan78 -- do you think we should proceed with the proposed approach or use a long-running branch and submit a huge PR in the end with all changes at once ?

pivotal-issuemaster commented:
@denalex Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=2634)!

Eulerizeit commented:
@d what exactly are you looking for? They seem to be following the practices described for all other work of this type.

pivotal-mike commented:
@denalex Have you considered writing this using the Postgres extensions framework?  That may be a more appropriate model for how this will interact with GPDB.  It also provides some more flexibility for further down the road if we want to convert this to using foreign data wrappers (FDW) when available.  It's a bit of work upfront but may put us in a better position later on.

hlinnaka commented on concourse/scripts/regression_tests_pxf.bash:
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash -l
```
Please hook the new regression tests into the top-level `installcheck-world` target, instead of adding new concourse jobs. That way, it will be run more frequently, and more easily, as part of everyone's routine development.

denalex replied:
@hlinnaka -- this was an original intention, however when the whole thing is done, this test will require PXF Java server setup in the test environment and perhaps even a reference implementation of Hadoop to be present as well. This will only be needed for PXF tests, so we decided to keep this separate from the very beginning, the same way gphdfs or gpcloud have their own separate CI test jobs. Makes sense ?

hlinnaka commented on configure.in:
```diff
@@ -731,6 +731,15 @@ AC_MSG_RESULT([checking whether to build with Greenplum Mapreduce... $enable_map
 AC_SUBST(enable_mapreduce)
 
 #
+# --enable-pxf enables Greenplum PXF support
```
We don't need an autoconf flag AFAICS. The extension doesn't depend on any external libraries or such, we can just always build it.

shivzone replied:
This flag was added to stay in lines with the gpcloud integration to allow the users the option

denalex replied:
this approach was also recommended by @pivotal-mike as well. Indeed, we do not have any external dependencies, other than libcurl, which is already configured. We can go either way, but having this flag provides the most flexibility, at least for now while the extension is being developed.

hlinnaka commented on gpAux/extensions/pxf/Makefile:
```diff
@@ -0,0 +1,41 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
What is the copyright / license status of this PR? If we later decide to relicense this, e.g. to submit this to the PostgreSQL community, is Pivotal free to do that? (Not that I think we're ever going to do it, but I'd like to be clear on the status of each part of the codebase.)

Assuming the license/copyright situation with this submission is the same as the rest of the codebase, I'd rather not include this blurp in every file. Most files in the repository don't carry these, the top-level README and LICENSE files cover the whole repository, unless otherwise stated.

shivzone replied:
I don't think the ASF licensing information isn't valid for this file or any other client/protocol code for PXF. The licensing info can be removed

denalex replied:
pinging @rvs to provide his feedback -- does code ported from Apache HAWQ (incubating) with no changes (or minimal changes) need to include Apache license header and what are the implications of including or not including it ?

Perhaps this makefile doesn't need it, but what about the rest of C code ?

rvs replied:
I don't quite get what's the problem with the licensing statement attached to the files. GPDB, as a body of work, is licensed under the Apache License. Period. That body of work happens to include parts that are ALSO available under PostgresSQL license. As such they can be used under the terms of both.

For this new functionality that is being developed the default assumption will be an Apache License regardless of whether the header says so or not. This is because all the new work flowing into GPDB is licensed under the Apache License (however because of the ICLA Pivotal can relicense it under BSD license if it sees fit and that way give it back to Postgres community).

So -- to recap -- regardless of the header the ORIGINAL new code is licensed under the Apache License anyway. As such header or no header should be up to the team based on considerations of backporting from other codebases, etc.

The EXISTING code has to preserve licensing headers. As such -- if there's any code lifted directly from Apache HAWQ (incubating) -- that code has to retain a header.

hlinnaka replied:
So the first question here is: Does this code submission include any 3rd party code? Is there any code included that was not written by a Pivotal employee, and thus cannot be relicensed by Pivotal in the future, if they'd wish to do so for some reason.

If the answer is "Yes", then we must retain the copyright notice and this license. And then we must also make sure that the NOTICE file that this refers to, contains accurate information about that.

If the answer is "No", then we (Pivotal) are free to do what we want. In that case, I think we should remove this license blurp, so that this is in line with the rest of the codebase. Having a separate license blurp where most of the codebase does not, would imply that this is somehow different from the rest of the codebase.

rvs replied:
I mostly agree, but I also want to make the following point explicit: even if the code is *partially* lifted from the ASF code base of HAWQ we'll have to keep the license header. I'll wait for @denalex to make an explicit statement of where the code originated.

schubert replied:
Right - so we'll wait for @denalex to comment on the code providence - and if YES @denalex then please also update `NOTICE` in this PR as well so we don't lose track of this important point.

denalex replied:
most source code files (.c) and some header files (.h) will be ported from Apache HAWQ. A few files might be renamed and refactored, but majority of actual code would come from Apache and we will preserve Apache header on these files. Makefiles, test cases and regression test scripts (.sql) will be new, so will not have the headers.

@rvs @hlinnaka @schubert -- how exactly should the NOTICE be updated ? I looked at that file (under root of the project: https://github.com/greenplum-db/gpdb/blob/master/NOTICE) and I assume I will need to add smth llike this, listing all the files with Apache license:

For the gpAux\extensions\pxf\pxfprotocol.c 
             gpAux\extensions\pxf\pxfprotocol.h 

components:

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

rvs replied:
Actually, instead of quoting the license header like that I'd recommend adding the top portion of the HAWQ's notice:

Apache HAWQ (incubating)
Copyright 2017 The Apache Software Foundation.

This product includes software developed at
The Apache Software Foundation (http://www.apache.org/).

You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0

denalex replied:
@rvs, so we do not need to list the files explicitly, just add whatever lines you provided to the NOTICE file ?

rvs replied:
Sorry for not being clear: you should also list files explicitly.

hlinnaka commented on gpAux/extensions/pxf/Makefile:
```diff
@@ -0,0 +1,41 @@
+MODULE_big = pxf
+OBJS       = src/pxfprotocol.o
+
+# Launch
```
What does "Launch" mean here?

denalex replied:
thanks, will remove

hlinnaka commented on gpAux/extensions/pxf/README.md:
```diff
@@ -0,0 +1,53 @@
+
+## Usage
+
+### Initialize and start GPDB cluster
```
I don't think we need to repeat the instructions on how to start up GPDB here.

denalex replied:
perhaps not, but was useful for newcomers to get started with setting up environment for PXF development

schubert replied:
@denalex it's markdown - can you hyperlink to the section where it's listed elsewhere? Otherwise we will end up with drift in instructions and that's no fun for anyone.

edespino replied:
These instructions are in the top-level README.md. We will update the PXF project's README.md to reference the same.

In fact, as it stands, the information in this README.md is out of order. One needs to build GPDB before instantiating the cluster.  It is correct in the top-level README.md.  Thank you for catching this.

hlinnaka commented on gpAux/extensions/pxf/regress/Makefile:
```diff
@@ -0,0 +1,16 @@
+top_builddir = $(shell cd ../../../.. && pwd)
```
The idiomatic place for regression tests in PostgreSQL extensions using PGXS is directly in sql/expected directories in the extension's top-level dir. In other words, remove the `regress` subdirectory, and move its contents one level up. Then you can hook the tests to the "installcheck" target by adding just `REGRESS = pxf` to the Makefile.

denalex replied:
right, that's how we did it originally, but then we moved into a separate directory to mimic what gpcloud extension did. @edespino -- would you like to comment on this ?

edespino replied:
As @denalex indicated, we followed the "gpcloud" extension implementation. The "gpcloud" extension uses "regress" to store "installcheck" files. We will adjust follow standard PostgreSQL extension layout conventions. Thank you for pointing this out.

hlinnaka replied:
Ok. You should know that I strongly disagreed with the way gpcloud was designed and packaged, but my objections were overruled. So if that's the model you follow, I'll continue to whine, and I assume I will continue to be overruled :-).

hlinnaka commented on gpAux/extensions/pxf/regress/sql/setup.sql:
```diff
@@ -0,0 +1,24 @@
+-- PXF Protocol/Formatters
+------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION pg_catalog.pxf_write() RETURNS integer
```
Please package the extension as a proper PostgreSQL/GPDB extension. We backported support for extensions into GPDB 5. See https://www.postgresql.org/docs/devel/static/extend-extensions.html for documentation on that, and existing upstream extensions for examples to follow (we haven't converted any extensions included in GPDB as extensions yet, but sooner or later that will happen).

denalex replied:
thanks for pointing this out. do you recommend doing this now and not when all other extensions will be converted ? if yes, @sansanichfb -- this might be a new independent task.

hlinnaka replied:
Yeah, I think we should package this as an extension from the get-go.

schubert replied:
+1 - yes let's bite the bullet now and make it an extension. At the very least we can have an exemplar for others to follow. I'm okay with that inconsistency since there's only one direction for everything else to go in.

hlinnaka commented on gpAux/extensions/pxf/src/pxfprotocol.c:
```diff
@@ -0,0 +1,108 @@
+Datum pxfprotocol_import(PG_FUNCTION_ARGS);
+Datum pxfprotocol_validate_urls(PG_FUNCTION_ARGS);
+
+/* include explicitly as cdbvars.h is not available when compiling as an extension */
```
This is really fragile. This will silently fail and crash if the GpId struct changes. Why can cdbvars.h not be included here?

denalex replied:
@hlinnaka -- if we strictly follow PGXS build guidelines and not include any makefiles explicitly, then an extension is supposed to be built on top of installed GPDB. We could not find cdbvars.h in GPDB install location -- while all other .h files were there, this one was not for some reason. It seems gpcloud ran into the same issue and also copied this struct. Is there any other function that will give us segindex at runtime ?

hlinnaka replied:
We'll need to fix that, then. It's just as bad for gpcloud. I don't know why cdbvars.h is not installed into include. We should install it, or move the declaration of GpIdentity elsewhere.

hlinnaka commented on gpAux/extensions/pxf/src/pxfprotocol.c:
```diff
@@ -0,0 +1,108 @@
+    int32 data_len = EXTPROTOCOL_GET_DATALEN(fcinfo);
+
+    /* produce a tuple */
+    sprintf(data_buf, "%d,hello world %d", GpIdentity.segindex, GpIdentity.segindex);
```
I know this is just mock code, but this a potential buffer overrun. It's straightforward to fix by using snprintf instead, so let's fix it.

denalex replied:
thanks, will change.

hlinnaka commented on gpAux/extensions/pxf/test/Makefile:
```diff
@@ -0,0 +1,7 @@
+subdir=gpAux/extensions/pxf
```
subdir should be `gpAux/extension/pxf/test`

denalex replied:
hmm, somehow it works with this value, will double check

hlinnaka replied:
$subdir happens to not be used in this file, but it would fail it if it was.

denalex commented on NOTICE:
```diff
@@ -2327,3 +2327,18 @@ documentation will at all times remain with copyright holders.
 
 The sample data used in the world database is Copyright Statistics
 Finland, http://www.stat.fi/worldinfigures.
+
```
@rvs -- updated NOTICE file, please review

rvs replied:
LGTM!

denalex commented:
@d @hlinnaka @pivotal-mike -- please review this PR again -- we made most of the changes requested, including:
- defined pxf as an extension
- removed /regress directory for tests
- removed GpIdentity re-declaration
- updated license files

We left in place --enable-pxf config flag to provide the most flexibility to exclude PXF from the builds, if needed. 

Getting the feedback and fixing the suggestions was a good exercise. The next PR will have a lot of real code and will provide end-to-end read from PXF use case (albeit without fragmenting / validation parts).

Please let us know if we can merge this PR into master now, so the scope for review of the next PR becomes smaller.

danielgustafsson commented on gpAux/extensions/pxf/src/pxfprotocol.c:
```diff
@@ -0,0 +1,99 @@
+#include "postgres.h"
+#include "fmgr.h"
+#include "cdb/cdbvars.h"
+#include "access/extprotocol.h"
```
Nitpick: please list these in alphabetical order. While the list is short (now), it makes it easier to avoid duplicate includes as it grows.

ascherbaum-pivotal commented on concourse/scripts/regression_tests_pxf.bash:
```diff
@@ -0,0 +1,65 @@
+	fi
+
+	if [ "$TARGET_OS" != "centos" -a "$TARGET_OS" != "sles" ]; then
+		echo "FATAL: TARGET_OS is set to an invalid value: $TARGET_OS"
```
s/invalid/not supported/

ascherbaum-pivotal commented on gpAux/extensions/pxf/src/pxfprotocol.c:
```diff
@@ -0,0 +1,99 @@
+Datum pxfprotocol_validate_urls(PG_FUNCTION_ARGS);
+
+typedef struct {
+    int32 row_count;
```
Given that we change other places to 64 bit right now - what are the odds that this can return more than 2B rows in the future?

denalex commented:
Thanks everyone for reviewing. It seems there is sufficient resistance for the merge of this PR in its current form. We will incorporate the most recent feedback into the ongoing work and will publish a new PR once we have real logic to read data from a running PXF server.
